doc/ntp/migration-to-chrony.md

```diff
@@ -0,0 +1,271 @@
+
+For this reason, it makes sense to migrate to chrony.
+
+## Migrating from ntpd to chrony in SONiC
+
+### Configuration
+
+In terms of SONiC configuration changes, there are no configuration changes
+required for migrating from ntpd to chrony. All of the configuration
+information passed in can be translated to chrony's syntax.
+
+For chrony's configuration file, there are differences in the configuration
+options that are available. The major ones of note are:
+* Chrony doesn't require listening on an interface to be able to send and
+  receive NTP packets on it.
+* When acting as a NTP server, chrony can only listen on one interface (or one
+  IPv4 and one IPv6 address), whereas ntpd can open any number of sockets
+  listening on port 123.
+* Chrony doesn't have a default panic threshold, whereas ntpd does by default.
+  A panic threshold means that if the time received via NTP is too different
+  than the system time (i.e. greater than the panic threshold), then the NTP
+  daemon will exit immediately instead of doing anything, and expect the system
+  administrator to first correct the system time to the actual time. For
+  SONiC's purposes, we do not want the panic threshold to be set. Chrony
+  doesn't set one, whereas ntpd does set a threshold of 1000 seconds by default
+  (which can be overridden).
+* ntpd's configuration specified what each subnet was allowed to do, whereas
+  chrony doesn't quite have that. This is partly because the configuration
+  control for chrony is on a different port entirely (UDP port 323 instead of
+  UDP port 123), this making it easier to be firewalled off and/or configured
+  separtely. In addition, chrony will default to using a client-server
+  relationship instead of symmetric relationship (where both sides will sync
+  time with each other), unless the `peer` keyword is used instead of `server`.
+* Chrony also allows storing the NTP servers in a separate file, making it
+  possible to reload the daemon and have it reread the servers instead of
+  restarting the whole daemon. At this time, this is not used in SONiC.
+* Chrony configuration file can specify `rtcsync` to tell the kernel that the
+  system time is now synchronized, and the kernel can then synchronize the
+  hardware clock/RTC with the system time. However, this would mean that if the
+  system time is significantly different from the actual time, then the
+  hardware clock/RTC will not get updated until the system time is synchronized
+  to the actual time, which may take months. As an alternative, chrony can
+  manage the hardware clock/RTC. With this, it will immediately update the
+  hardware clock/RTC with the actual time, while the system time is gradually
+  slewed. This will be the configuration chosen for SONiC.
+
+### Monitoring
+
+For the purpose of making time synchronization issues more visible, a Monit
+check will be added to verify that the time is currently synchronized to one or
```

Comment:
It would be good if the minimum number of active servers could be defined so that the synchronization is considered correct. For high requirements, 3 servers should be the minimum requirement.

It would also be good if the details of the time synchronization station (e.g. offset, jitter, delay, count) could be monitored from the network. As far as I know there is currently no Prometheus exporter included in SONiC, SNMP would probably be the only option here.

Reply:
For the first part, while specifying 3 servers is required for a high level of time accuracy, I'd rather not make it a minimum requirement within SONiC, only because if it is used in an environment that either doesn't need a high level of time accuracy or doesn't have 3 servers available, then that requirement won't be met.

For the second part, I agree on exposing the metrics somehow. I think the current standard we have is to publish the data into STATE_DB, which should make it easier to get exported elsewhere. However, this would involve having (at minimum) another daemon polling chrony frequently, and publishing the data into STATE_DB, and the scope of this effort is already large enough. Can you open an enhancement request for this?

Reply:
About the minimum servers required: I fully agree that there are environments where having accurate time is not that important. My recommendation was to make the minimum numbers of servers "defineable", then its up to the operator what is needed.

Regarding metrics topic:
Yes of course, see https://github.com/sonic-net/SONiC/issues/1857